text/0000-abstract-prerender.md

```diff
@@ -0,0 +1,84 @@
+- Start Date: 2021-05-29
+- RFC PR: 
+- Svelte Issue: 
+
+# Abstract pre-render
+
+## Summary
+
+Adding support for prerendered HTML files which look for their page store data at client-side runtime (what I'd call "abstract"), instead of using the snapshot of such data as it was at build time (what I'd call "concrete").
+
+## Motivation
+
+This will allow us to pre-render every page, even the ones that do something with the request-specific page store data in `onMount`. It is possible to do this today with no changes to SvelteKit itself, if you have a web server that can read and transform (interpolating values) into the html artifacts for each request. But doing that is messy.
+
+There's a lot more detail about motivation in the initial comment of https://github.com/sveltejs/kit/issues/1533.
+
+## Detailed design
+
+### Technical Background
+
+Right now, a pre-rendered page has code like this to hydrate (this is for a parameterized route):
+
+    page: {
+      host: location.host, // TODO this is redundant
+      path: "/my/page/123",
+      query: new URLSearchParams("foo=bar"),
+      params: {"id": "123"}
+    }
+
+Aside from the "host", those are all the concrete, specific values as they were at build time, baked into the page as literals. But one might want to use the prerendered HTML artifact in an abstract way, for variable and wildcard request parameters, in which case you'd want something more like this:
+
+    page: {
+      host: location.host,
+      path: location.pathname,
+      query: new URLSearchParams(location.search),
+      params: parsePathParams(location.pathname, /^\/my\/page\/(\w)+$/, ["id"]),
+    }
+
+Where `parsePathParams` is a simple function that gives you an object of path params.
+
+### Implementation
+
+The current state of the code is here:
+
+https://github.com/sveltejs/kit/blob/a66affec4168fae053fb6349805b60028306abcf/packages/kit/src/runtime/server/page/render.js#L136
+
+The changes would be:
+
+* For host, remove the TODO comment and the ternary, and just always use `location.host`
+* For path, use `location.pathname`
+* For query, use `location.search`
+* Params is the hardest one, we'd need to insert the route regex and the positional list of parameter names into a simple `parsePathParams` function, possibly just inlined as an arrow function
```

Comment:
I'd proposed moving the parsing from built time to the runtime in https://github.com/sveltejs/rfcs/pull/36

Reply:
Thanks, hadn't seen that before, let me look.